raft_topology: fencing and global_token_metadata_barrier

gusev-p opened the pull request:
This is the initial implementation of [this spec](https://docs.google.com/document/d/1X6pARlxOy6KRQ32JN8yiGsnWA9Dwqnhtk7kMDo8m9pI/edit).

* the topology version (int64) was introduced, it's stored in topology table and updated through RAFT at the relevant stages of the topology change algorithm;
* when the version is incremented, a `barrier_and_drain` command is sent to all the nodes in the cluster, if some node is unavailable we fail and retry indefinitely;
* the `barrier_and_drain` handler first issues a `raft_read_barrier()` to obtain the latest topology, and  then waits until all requests using previous versions are finished; if this round of RPCs is finished the topology change coordinator can be sure that there are no requests inflight using previous versions and such requests can't appear in the future.
* after `barrier_and_drain` the topology change coordinator issues the `fence` command, it stores the current version in local table as `fence_version` and blocks requests with older versions by throwing `stale_topology_exception`; if a request with older version was started before the fence, its reply will also be fenced.
* the fencing part of the PR is for the future, when we relax the requirement that all nodes are available during topology change; it should protect the cluster from requests with stale topology from the nodes which was unavailable during topology change and which was not reached by the `barrier_and_drain()` command;
* currently, fencing is implemented for `mutation` and `read` RPCs, other RPCs will be handled in the follow-ups; since currently all nodes are supposed to be alive the missing parts of the fencing doesn't break correctness;
* along with fencing, the spec above also describes error handling, isolation and `--ignore_dead_nodes` parameter handling, these will be also added later; [this ticket](https://github.com/scylladb/scylladb/issues/14070) contains all that remains to be done;
* we don't worry about compatibility when we change topology table schema or `raft_topology_cmd_handler` RPC method signature since the raft topology code is currently hidden by `--experimental raft` flag and is not accessible to the users. Compatibility is maintained for other affected RPCs (mutation, read) - the new `fencing_token` parameter is `rpc::optional`, we skip the fencing check if it's not present.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1242/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1244/

avikivity commented:
> This is the initial implementation of [this spec](https://docs.google.com/document/d/1X6pARlxOy6KRQ32JN8yiGsnWA9Dwqnhtk7kMDo8m9pI/edit).


Let's port it to a .md and commit it. I hope some tooling can help with that,

> 
> There is no error handling, isolation and `--ignore_dead_nodes` handling, these will be added later.
> 
> Fencing is only added to `mutation` and `read` RPCs, other RPCs will be handled in the follow-ups.

What's the impact of the missing functionality? Will it cause incorrect behavior? Or are we protected by the timeouts (as far as timeouts are protection)?

avikivity commented:
I see the spec should be easy to port.

Please also expand on the cover letter a little.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -257,6 +257,7 @@ schema_ptr system_keyspace::topology() {
             .with_column("shard_count", int32_type)
             .with_column("ignore_msb", int32_type)
             .with_column("new_cdc_generation_data_uuid", uuid_type)
+            .with_column("version", long_type, column_kind::static_column)
```
How does this survive upgrades?

Given this is a raft table, do we need to coordinate it with other nodes? 

Note that 5.3 will be branched soon, so we must be prepared for 5.3->master upgrades.

gleb-cloudius replied:
The table is behind experimental. Should not be an issue.

avikivity commented on locator/token_metadata.cc:
```diff
@@ -85,6 +85,8 @@ class token_metadata_impl final {
     static thread_local long _static_ring_version;
 
+    token_metadata::version_t _version = 0;
+
```
Should this be std::optional<> to distinguish between "we had a version 0" and "we did not read any version"?

kbr-scylla replied:
> In particular, in RPCs, if fencing_token.version == 0 we know the caller is not yet aware of versions/fencing

Wouldn't we receive `nullopt` in that case?

avikivity commented on locator/token_metadata.cc:
```diff
@@ -310,6 +312,22 @@ class token_metadata_impl final {
+                       "new version {}, previous version {}", version, _version));
+        }
+        _version = version;
+    }
```
Q (not directly related):

How are we protected from this node losing leadership, some other node advancing the topology, and then this node regaining leadership and committing a now-outdated _version?

avikivity commented on idl/storage_service.idl.hh:
```diff
@@ -10,10 +10,12 @@ namespace service {
   struct raft_topology_cmd {
       enum class command: uint8_t {
           barrier,
+          barrier_and_drain,
```
Breaks ABI compatibility. Add new enums at the end.

Must protect with a cluster feature to avoid use before it's ready.

Or perhaps we can sweep all this under the carpet since it's still experimental.

gleb-cloudius replied:
It is experimental. We can change it freely at this point.

avikivity commented on idl/storage_service.idl.hh:
```diff
@@ -10,10 +10,12 @@ namespace service {
           fence_old_reads
       };
       service::raft_topology_cmd::command cmd;
+      std::variant<std::monostate, service::topology::version_t> param;
```
I think a nicer way is to combine the enum and the parameters.

```c++
struct barrier {};
struct barrier_and_drain { version_t version };
...
struct fence_old_reads {};

std::variant<barrier, barrier_and_drain, ..., fence_old_reads> payload;
```

avikivity replied:
(no way to use the wrong parameters for your enum)

avikivity commented on idl/storage_proxy.idl.hh:
```diff
@@ -21,8 +21,9 @@
+#include "idl/storage_service.idl.hh"
 
-verb [[with_client_info, with_timeout, one_way]] mutation (frozen_mutation fm, inet_address_vector_replica_set forward, gms::inet_address reply_to, unsigned shard, uint64_t response_id, std::optional<tracing::trace_info> trace_info [[version 1.3.0]], db::per_partition_rate_limit::info rate_limit_info [[version 5.1.0]]);
+verb [[with_client_info, with_timeout, one_way]] mutation (frozen_mutation fm, inet_address_vector_replica_set forward, gms::inet_address reply_to, unsigned shard, uint64_t response_id, std::optional<tracing::trace_info> trace_info [[version 1.3.0]], db::per_partition_rate_limit::info rate_limit_info [[version 5.1.0]], service::fencing_token fence [[version 5.3.0]]);
```
Won't make it into 5.3.

gleb-cloudius replied:
It tells idl compiler that the parameter should be rpc::optional.

gleb-cloudius replied:
It is a documentation thing. Yes.

avikivity replied:
@bhalevy also plans to use them to generate IDL signatures so that we can mechanically check compatibility.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -182,7 +182,7 @@ class storage_proxy::remote {
         ser::storage_proxy_rpc_verbs::register_counter_mutation(&_ms, std::bind_front(&remote::handle_counter_mutation, this));
         ser::storage_proxy_rpc_verbs::register_mutation(&_ms, std::bind_front(&remote::receive_mutation_handler, this, sp->_write_smp_service_group));
-        ser::storage_proxy_rpc_verbs::register_hint_mutation(&_ms, [this, sp] <typename... Args>(Args&&... args) { return receive_mutation_handler(sp->_hints_write_smp_service_group, std::forward<Args>(args)..., std::monostate()); });
+        ser::storage_proxy_rpc_verbs::register_hint_mutation(&_ms, [this, sp] <typename... Args>(Args&&... args) { return receive_mutation_handler(sp->_hints_write_smp_service_group, std::forward<Args>(args)..., std::monostate(), std::nullopt); });
```
`the_type()` to avoid people having to lookup.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -411,7 +412,7 @@ class storage_proxy::remote {
             auto schema_version, auto in, inet_address_vector_replica_set forward, gms::inet_address reply_to,
             unsigned shard, storage_proxy::response_id_type response_id, std::optional<tracing::trace_info> trace_info,
-            auto&& apply_fn1, auto&& forward_fn1) {
+            fencing_token fence, auto&& apply_fn1, auto&& forward_fn1) {
```
I hate this function

gleb-cloudius replied:
Why? Not enough parameters?

avikivity replied:
Too many autos

gleb-cloudius replied:
It is possible to specify real types here, but the function definition will be even longer.

avikivity replied:
Even worse, someone might be able to understand what's going on

avikivity commented on service/storage_proxy.cc:
```diff
@@ -446,51 +447,67 @@ class storage_proxy::remote {
+                        co_await apply_fn(p, trace_state_ptr, std::move(s), m, timeout);
+
+                        if (auto stale = try_apply_fence(fence); stale) {
+                            throw std::move(*stale);
```
Why is this repeated?

avikivity replied:
Ah, you say why. I think it deserves a comment here.

gleb-cloudius replied:
The idea is to send failure to the request coordinator so it knows that it may have written using stale topology. Not sure this is needed if we drain without a timeout.

avikivity replied:
But why try_apply_fence twice?

gusev-p replied:
> Not sure this is needed if we drain without a timeout.

Currently it's not needed since we `barrier_and_drain` all replicas. It's included to resolve @kbr-scylla case from the link above, if in the future we would drain only some affected replicas.

gleb-cloudius replied:
> But why try_apply_fence twice?

We can fence only once after the op, but then you may do unnecessary work. It is like optimistic locking.

avikivity replied:
How can you fence after the op, it if's a write? Isn't it to late?

Since we grab the reference to the topology, isn't it enough to check it at the beginning? The fence rpc will wait until we release it.

gleb-cloudius replied:
You need either wait in the fence rpc for old version to complete or re-check after the op and drop a reply if fence changed in between. Here second option is implemented. Is more live (you do not need to wait while changing topology) and waiting for replica writes will be needed only before running cleanup.

avikivity commented on utils/tuple_utils.hh:
```diff
@@ -0,0 +1,27 @@
+#include <functional>
```
License/copyright.

avikivity commented on utils/tuple_utils.hh:
```diff
@@ -0,0 +1,27 @@
+    // ResultTuple mirrors SourceTuple, with the inclusion of an extra element of type NewItem.
+    // SourceTuple and ResultTuple can be arbitrary specialisations of std::tuple or rpc::tuple.
+    template<typename ResultTuple, typename SourceTuple, typename NewItem>
+    requires (std::tuple_size_v<SourceTuple> + 1 == std::tuple_size_v<ResultTuple>)
```
Let's require that they're tuples too.

avikivity commented on utils/tuple_utils.hh:
```diff
@@ -0,0 +1,27 @@
+        constexpr auto result_size = std::tuple_size_v<ResultTuple>;
+        constexpr auto item_index = std::invoke([]<std::size_t... Is>(std::index_sequence<Is...>) {
+            auto result = result_size;
+            (void) ((std::is_same_v<std::tuple_element_t<Is, ResultTuple>, NewItem> && (result = Is, true)) || ...);
```
This depends on types not repeating, which is fragile for a general utility, but ok.

avikivity commented on utils/tuple_utils.hh:
```diff
@@ -0,0 +1,27 @@
+                std::get<Right + item_index>(std::forward<SourceTuple>(source_tuple))...
+            };
+        }, std::make_index_sequence<item_index>(), std::make_index_sequence<source_size - item_index>());
+    }
```
These C++ template games are contortions, but I love them.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -604,102 +604,110 @@ class storage_proxy::remote {
-        future<rpc::tuple<foreign_ptr<lw_shared_ptr<query::result>>, cache_temperature>> f = co_await coroutine::as_future(p->query_result_local(std::move(s), cmd, std::move(pr2.first), opts, trace_state_ptr, timeout, rate_limit_info));
-        tracing::trace(trace_state_ptr, "read_data handling is done, sending a response to /{}", src_ip);
-        co_return co_await encode_replica_exception_for_rpc<rpc::tuple<foreign_ptr<lw_shared_ptr<query::result>>, cache_temperature, replica::exception_variant>>(p->features(), std::move(f));
+        auto f = co_await coroutine::as_future(std::invoke([&]() {
```
Did this introduce a new allocation?

Better to move the template to the calling scope if so.

gusev-p replied:
There were the same `coroutine::as_future` in the previous version. What allocation do you mean?

avikivity replied:
Every coroutine introduces an allocation. But there's net zero coroutine changes, so we're fine.

avikivity commented on idl/storage_service.idl.hh:
```diff
@@ -16,7 +16,6 @@ struct raft_topology_cmd {
         barrier,
         barrier_and_drain,
         stream_ranges,
-        fence_old_reads,
```
Can't be removed, if we maintain compatibility.

gusev-p replied:
It's not published yet, under experimental flag.

gleb-cloudius replied:
We do not at this point.

avikivity commented:
What about tests?

In particular I want to see inducted violation of fencing.

Note sure how to do that. Perhaps isolate a node from all-but-one of its peers (making sure the one isn't the topology leader), and see that its ops generate topology exceptions.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -158,6 +158,18 @@ class storage_proxy::remote {
 
+    std::optional<replica::stale_topology_exception> try_apply_fence(fencing_token token) const noexcept {
+        // topology_version == 0 means the caller is not aware about the fencing yet
+        const auto fence_version = _sp._shared_token_metadata.get_fence_version();
```
Shouldn't this be done on the op's captured effective_replication_map_ptr?

Otherwise, by the time we capture it, it may change. Or if we captured it before, we may have captured an old one. Maybe this should be a method of effective_replication_map.

gleb-cloudius replied:
The op should use only one token metadata version on a coordinator. But on a replica we need to make sure that by the time reply is sent the version the coordinator uses is still not fenced.

avikivity replied:
We should be able to run with different-but-compatible topology versions.

gleb-cloudius commented on locator/token_metadata.cc:
```diff
@@ -1282,6 +1300,15 @@ token_metadata::invalidate_cached_rings() {
     _impl->invalidate_cached_rings();
 }
 
+auto
```
Why not specify ret type here instead of in the end?

gleb-cloudius commented on locator/token_metadata.cc:
```diff
@@ -1308,12 +1312,24 @@ void
 void shared_token_metadata::set(mutable_token_metadata_ptr tmptr) noexcept {
     if (_shared->get_ring_version() >= tmptr->get_ring_version()) {
-        on_internal_error(tlogger, format("shared_token_metadata: must not set non-increasing version: {} -> {}", _shared->get_ring_version(), tmptr->get_ring_version()));
+        on_internal_error(tlogger, format("shared_token_metadata: must not set non-increasing ring_version: {} -> {}", _shared->get_ring_version(), tmptr->get_ring_version()));
```
Is this belongs to this patch or previous one when new kind of version was introduced?

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -4845,6 +4845,24 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                 break;
+                case raft_topology_cmd::command::barrier_and_drain:
+                    co_await container().invoke_on_all([param = cmd.param] (storage_service& ss) -> future<> {
+                        const auto requested_version = holds_alternative<topology::version_t>(param)
```
IMO there is no need to pass the version we want to drain explicitly. If we are here it means current metadata version is up-to-date and we need to wait for all earlier versions to complete.

gleb-cloudius replied:
You first store the same state in to different places (rpc call and topology) and then you try to make sure they do not go out of sync while you have no a way to recover if they do. What is the point? The thing is all other rpc calls here already use the way I described. I think we should continue at least for consistency.

gleb-cloudius replied:
OTOH we need to protect from commands been re-ordered. We already have a protection from executing a command from outdated coordinator by having the term in the cmd and checking it before executing a command. What we need is to add a counter that is monotonically increased for each term and remember the id of the last command. The cmd can be executed only if it's id is higher than the previous one otherwise it is dropped. This way we can fully rely on a leader state and a follower state to be in sync and to not add different checks in each and every cmd.

gusev-p replied:
ok

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -5000,13 +5002,27 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                     result.status = raft_topology_cmd_result::command_status::success;
                 break;
+                case raft_topology_cmd::command::fence: {
+                    const auto version = get<topology::version_t>(cmd.param);
```
Here I also think that the version parameter can be implicit and taken from the local topology object.

gleb-cloudius commented on replica/exceptions.hh:
```diff
@@ -43,10 +43,34 @@ public:
+class stale_topology_exception final : public replica_exception {
+    int64_t _caller_version;
+    int64_t _callee_fence_version;
+    seastar::sstring _message;
```
You already have message string in `std::exception` you inherit from.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -446,51 +447,67 @@ class storage_proxy::remote {
+                        co_await apply_fn(p, trace_state_ptr, std::move(s), m, timeout);
+
+                        if (auto stale = try_apply_fence(fence); stale) {
+                            throw std::move(*stale);
```
This should never be needed in case drain is done without timeout, no?

gusev-p replied:
it would be needed if we drained only part of the replicas, it's demonstrated by the @kbr-scylla example here https://docs.google.com/document/d/1X6pARlxOy6KRQ32JN8yiGsnWA9Dwqnhtk7kMDo8m9pI/edit?disco=AAAAuXv0tVk

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -446,51 +447,67 @@ class storage_proxy::remote {
+            errors.local = std::move(*stale);
+        }
+
+        if (errors.count == 0) {
```
It is better to leave the indentation and re-indent in a separate patch.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -539,13 +559,14 @@ class storage_proxy::remote {
         auto schema_version = decision.update.schema_version();
         return handle_write(src_addr, t, schema_version, std::move(decision), std::move(forward), reply_to, shard,
                 response_id, trace_info,
+                fencing_token{},
```
Why don't we pass it for paxos?

gusev-p replied:
skipped paxos RPCs in this PR, to restrict the scope of the PR/get early feedback

avikivity replied:
> v2:
> 
> * addressed all the comments

That's really making me re-review everything. With a real changelog I can focus on the more important changes.


> * converted the spec to .md
> * added description to the PR
> 
> @avikivity @gleb-cloudius please re-review

avikivity replied:
Don't know why this got threaded here.

gusev-p replied:
> That's really making me re-review everything. With a real changelog I can focus on the more important changes.

* `fence_version` moved from `topology` table to `scylla_local`;
* `docs/dev/raft-topology.md` is added;
* changed version in RPC description to `[[version 5.4.0]]`;
* removed `param` from raft commands, use current version after `read_barrier`;
* added `cmd_index` to raft commands to protect agains commands reordering;
* `try_apply_fence` renamed to `apply_fence`, added overload with `future` parameter;
* specific type instead of `std::monostate()` in `register_hint_mutation`;
* use `apply_fence` with future overload to fence read and write requests;
* added fencing to `send_to_live_endpoints`, handle `stale_topology_exception`;

avikivity replied:
Thanks.

Loading from scylla_local makes sense (and answers my question about it being raft controlled), but we should make an effort to sync with the cluster before accepting ops, to avoid needless fencing.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -1451,7 +1474,9 @@ class abstract_write_response_handler : public seastar::enable_shared_from_this<
-        return _mutation_holder->apply_remotely(*_proxy, ep, std::move(forward), response_id, timeout, std::move(tr_state), _rate_limit_info);
+        return _mutation_holder->apply_remotely(*_proxy, ep, std::move(forward),
+            response_id, timeout, std::move(tr_state), _rate_limit_info,
+            {_effective_replication_map_ptr->get_token_metadata().get_version()});
```
Shouldn't we also do fencing on local writes? If we do drain with timeout a local write may complete after the fence moved.

gusev-p commented on idl/storage_service.idl.hh:
```diff
@@ -7,37 +7,37 @@
-  };
+struct fencing_token {
+    service::topology::version_t topology_version;
+};
```
ok, for next time

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -4674,7 +4696,9 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
-            return _proxy->remote().send_read_mutation_data(netw::messaging_service::msg_addr{ep, 0}, timeout, _trace_state, *cmd, _partition_range);
+            return _proxy->remote().send_read_mutation_data(netw::messaging_service::msg_addr{ep, 0}, timeout,
+                _trace_state, *cmd, _partition_range,
+                {_effective_replication_map_ptr->get_token_metadata().get_version()});
```
Same about local read as I wrote about write.

gusev-p replied:
Yes, will fix.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1087,7 +1116,8 @@ class topology_coordinator {
                                 .del("topology_request")
-                                .set_transition_state(topology::transition_state::write_both_read_old);
+                                .set_transition_state(topology::transition_state::write_both_read_old)
+                                .set_version(_topo_sm._topology.version + 1);;
```
double semicolon.

gusev-p replied:
noted, will fix

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1098,7 +1128,8 @@ class topology_coordinator {
                                 .del("topology_request")
-                                .set_transition_state(topology::transition_state::write_both_read_old);
+                                .set_transition_state(topology::transition_state::write_both_read_old)
+                                .set_version(_topo_sm._topology.version + 1);;
```
double semicolon.

gusev-p replied:
noted, will fix

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -258,6 +258,7 @@ schema_ptr system_keyspace::topology() {
             .with_column("ignore_msb", int32_type)
             .with_column("new_cdc_generation_data_uuid", uuid_type)
             .with_column("version", long_type, column_kind::static_column)
+            .with_column("fence_version", long_type, column_kind::static_column)
```
> > This table is updated with raft and this field is not.
> 
> Does this have any technical ramifications?

It has logical ramifications. A state updated through raft and not raft will be mixed. We need at least a comment, but I prefer to have it in separate tables all together.

gusev-p commented:
> What about tests?
> 
> In particular I want to see inducted violation of fencing.
> 
> Note sure how to do that. Perhaps isolate a node from all-but-one of its peers (making sure the one isn't the topology leader), and see that its ops generate topology exceptions.

Currently topology change coordinator fail/retry if it can't contact a node. We could create such test later when we implement `--ignore-dead-nodes` parameter.

gusev-p commented:
v2:
  * addressed all the comments
  * converted the spec to .md
  * added description to the PR

@avikivity @gleb-cloudius please re-review

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1327/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1349/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1351/

avikivity commented:
> * we don't handle upgrades when we change topology table schema or RPC methods signatures since the raft topology code is currently hidden by `--experimental raft` flag and is not accessible to the users.

rpc signatures must still be compatible, since those rpc verbs are available without --experimental. On the day that we make this available, it will need to support upgrades.

The schema changes on a table that is hidden by --experimental, so we can be more relaxed.

I believe the code is okay, but please double check, and update the wording here.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -3705,6 +3706,20 @@ future<service::topology> system_keyspace::load_topology_state() {
+    co_await execute_cql(
+        format("INSERT INTO system.{} (key, fence_version) VALUES (?, ?)", TOPOLOGY),
+        sstring(TOPOLOGY), value);
+    co_await force_blocking_flush(TOPOLOGY);
```
Is it okay to update a raft-controlled table locally? I'm sure it isn't.

gusev-p replied:
moved `fence_version` to `scylla_local`

avikivity commented on service/storage_proxy.cc:
```diff
@@ -582,102 +582,110 @@ class storage_proxy::remote {
+                os << "read_digest";
+                break;
+        }
+        return os;
```
New code should use fmt formatters. Don't make more work for @tchaikov .

gusev-p replied:
should this `::format("{}", read_verb::read_data)` just work or should I implement something?

avikivity replied:
See 3928a9a4e9bb3e0f6caea09bba9faee275044d01 for an example (and https://fmt.dev/latest/api.html#formatting-user-defined-types for the theory)

gusev-p replied:
This `read_verb` is a local, ad-hoc, private enum, nested in some class, I can't partially specialise a class inside another class and I don't want to make this enum more public than it's needed.

@tchaikov What's the problem with `operator<<`? I would prefer to turn it to just local lambda `read_verb->string_view` if it's so bad.

gusev-p commented on service/storage_service.hh:
```diff
@@ -781,8 +781,12 @@ private:
+    struct {
+        raft::term_t term{0};
+        size_t last_index{0};
+    } _raft_topology_cmd_handler_state{};
```
one `{}` is enough - either at `_raft_topology_cmd_handler_state` or at `term` and `last_index` both produce zero initialisation, but is it really important?

gleb-cloudius replied:
> but is it really important?

Of course not. It is like complaining about double ';'. If you need to produce a new version of the series anyway drop one, but otherwise no need to bother.

gusev-p replied:
`nit:` tag could help to make this intention clear

gusev-p commented on service/storage_service.cc:
```diff
@@ -4869,6 +4874,18 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
            co_return result;
         }
+
+        {
```
to restrict the scope of the `state` local variable

gusev-p replied:
We don't need to do it for each and every variable, what a strange desire. The point is to limit a block of code that performs one logically isolated task, so that it does not leave behind garbage - used and no longer needed variables. If the containing function is large, these variables may accidentally conflict with some other variables below and cause unnecessary questions. This is the same reason why `if(type var = init(); cond(var)) { use(var); }` syntax was introduced, but it's not convenient to use it here.

gleb-cloudius replied:
> We don't need to do it for each and every variable, what a strange desire
Why not? Why this one is special?

> The point is to limit a block of code that performs one logically isolated task

If the work worth to be isolated it should be in a function.

> cause unnecessary questions

What kind of questions? There will be an error about variable re-declaration. It may mean that the name of the variable was not descriptive enough.

About the 'if' you do not introduce a scope in the middle of nowhere. You have a scope and it is of course fine to limit a variable to it, but adding scopes in random places does not help. On the contrary.

gleb-cloudius replied:
> > If the work worth to be isolated it should be in a function.
> 
> No, is depends. The function by itself adds cognitive load, if the code is used only in one place it's more often doesn't deserve to be extracted.

Good named function reduces cognitive load. The code that is used in one place does not deserve global function, I agree. But C++ have lambdas to simulate local functions easily.

> 
> > What kind of questions? There will be an error about variable re-declaration.
> 
> Why am I getting this error?

If you declare the same variable twice you get an error about re-declaration.

> 
> > It may mean that the name of the variable was not descriptive enough.
> 
> Braces allows use not to invent artificial names for things that are basically the same, but used in slightly different contexts in the same function. Here's the typical example: https://github.com/gusev-p/scylla/blob/095f35a47d4dafe5f08f8a76c1fb49b16c2b7592/test/boost/token_metadata_test.cc#L55
> 
> We have several cases, each case lives in its own block, we don't need to invent a name for `erm` and can use the same code structure for both of them.
> 
> I don't see what's wrong with these braces.

I do not like them for three reasons: take vertical space, take horizontal space, make one think why they are there (cognitive load at you put it). For cases where you need to limit a guard lifetime or want to destroy unneeded large object in a sleeping function such artificial scope has a lot of sense, so when I see it I start looking for those reasons.

Anyway this is a taste thing. No need to block the series because of it.

kbr-scylla replied:
To me these local scopes reduce cognitive load, because they make lifetimes of variables smaller. I know the variables declared in the scope won't be used later.

gleb-cloudius replied:
If a function is so long that it matters may be it should be split into several functions. But we discussed that already. I do not think we can come up with new arguments.

gusev-p commented:
rebased on master

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1438/

gusev-p commented:
merge/build fix

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1440/

kbr-scylla commented on service/storage_proxy.cc:
```diff
@@ -2812,6 +2812,32 @@ storage_proxy::mutate_hint(const schema_ptr& s, const frozen_mutation& m, tracin
+    if (!token || token.topology_version >= fence_version) {
+        return std::nullopt;
+    }
+    slogger.warn("Stale topology detected, request has been fenced out, "
```
This will cause a huge spam of WARN-level logs if there's a high throughput in writes and we hit this.

gusev-p replied:
It's important event, deserves a `WARN` IMHO. I can wrap it in a some rate limiter, I remember it was done somewhere, but can't find an example...

kbr-scylla replied:
Rate limiter sounds good
example usage: https://github.com/scylladb/scylladb/blob/af65d5a1e8e61565da6ef60ae7a4a8dab675e0f6/service/raft/raft_group_registry.cc#L473

gusev-p replied:
thanks!

kbr-scylla commented on service/storage_proxy.cc:
```diff
@@ -170,7 +170,7 @@ class storage_proxy::remote {
         ser::storage_proxy_rpc_verbs::register_counter_mutation(&_ms, std::bind_front(&remote::handle_counter_mutation, this));
         ser::storage_proxy_rpc_verbs::register_mutation(&_ms, std::bind_front(&remote::receive_mutation_handler, this, sp->_write_smp_service_group));
-        ser::storage_proxy_rpc_verbs::register_hint_mutation(&_ms, [this, sp] <typename... Args>(Args&&... args) { return receive_mutation_handler(sp->_hints_write_smp_service_group, std::forward<Args>(args)..., std::monostate()); });
+        ser::storage_proxy_rpc_verbs::register_hint_mutation(&_ms, [this, sp] <typename... Args>(Args&&... args) { return receive_mutation_handler(sp->_hints_write_smp_service_group, std::forward<Args>(args)..., std::monostate(), rpc::optional<fencing_token>{}); });
```
FIXME?

gusev-p replied:
There are todo's in my personal todo list to fix this and other RPCs, they will be done right after this PR, fixme's in code won't help much.

kbr-scylla replied:
personal todo lists have a small bus factor

gusev-p replied:
I can create a ticket with bullets of what needs to be done.

gusev-p replied:
https://github.com/scylladb/scylladb/issues/14070

kbr-scylla commented on service/storage_proxy.cc:
```diff
@@ -498,22 +505,25 @@ class storage_proxy::remote {
-                        clock_type::time_point timeout) {
-                    return p->mutate_locally(std::move(s), m, std::move(tr_state), db::commitlog::force_sync::no, timeout, smp_grp, rate_limit_info);
+                        clock_type::time_point timeout, fencing_token fence) {
+                    return p->apply_fence(p->mutate_locally(std::move(s), m, std::move(tr_state), db::commitlog::force_sync::no, timeout, smp_grp, rate_limit_info), fence);
```
I don't understand this. We apply the write to the replica - but then lie to the coordinator that it was fenced? What's the point if it wasn't really fenced, we actually wrote it?

Checking the version at any moment after the data was written doesn't make sense to me. And in the commit message you write:
> apply_fence after the local write guarantees that no mutations succeed on coordinators if the fence version has been updated on the replica during the write.

We may reply success, but then our fence version changes while the reply message is in transit to the coordinator. What then? To an external observer, this is equivalent to the fencing version changing "during the write". Basically "duration of the write" is between coordinator generating the write and getting a response. Not the duration of `mutate_locally` on the replica.

gleb-cloudius replied:
> 
> We may reply success, but then our fence version changes while the reply message is in transit to the coordinator. What then? To an external observer, this is equivalent to the fencing version changing "during the write". Basically "duration of the write" is between coordinator generating the write and getting a response. Not the duration of `mutate_locally` on the replica.

If the version changes after reply it means it changed after the data was written and after the barrier done by the topology coordinator completed and hence it means that the data will be included in streaming and thus it is valid to reply with success to the write.

kbr-scylla replied:
Ok, in other words
- if the version did not increase after the write was done on the replica, then we can be sure that this write will be included in streaming. So we can respond with ACK
- if we see that the version increased after the write was done on the replica, then we're not sure if this write will be included in streaming. Therefore we shouldn't ACK even though we successfully persisted the write.

In the second case it doesn't matter that the version may increase while the ACK is in transit. The write will persist.

So it makes sense. Thanks.

gusev-p commented on service/storage_service.cc:
```diff
@@ -1777,6 +1777,8 @@ future<> storage_service::join_token_ring(cdc::generation_service& cdc_gen_servi
 
         co_await update_topology_with_local_metadata(*raft_server);
 
+        co_await update_fence_version(co_await _sys_ks.local().get_topology_fence_version());
```
> Won't this race with update_fence_version coming from coordinator commands, which might be arriving at this point?

I was under impression that we don't accept external communication until that point

> Also on restart I think we should do this earlier - because IIRC at this point we might already be performing data path ops.

Indeed, `messaging_service::start_listen` is called before `storage_service::join_cluster`, so we should call this somewhere before.

gleb-cloudius replied:
> > I was under impression that we don't accept external communication until that point
> 
> How would we join group 0 and perform a read barrier?

May patch will create group0 server before any message is received to solve the race where messages that need group0 started to arrive before it was created. Doing anything through this server will be possible only after messaging service starts listening.

gusev-p commented on service/storage_service.cc:
```diff
@@ -5198,13 +5200,27 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                 break;
+                case raft_topology_cmd::command::fence: {
+                    const auto version = _topology_state_machine._topology.version;
+                    co_await update_fence_version(version);
```
No, it should be run sequentially by the topology change coordinator.

kbr-scylla replied:
What if there's coordinator failover?

gusev-p replied:
It seems nothing prevents a race:
* coordinator A with fence_version=1 reached `update_fence_version`
* coordinator B with fence_version=2 came
* raft already updated its term, so it's matched
* coordinator B finishes `update_fence_version`, checks pass since fence_version=2 > 1 - the previous version
* coordinator B finishes `_sys_ks.local().update_topology_fence_version`;
* coordinator A finishes `_sys_ks.local().update_topology_fence_version`, overwriting fence_version back to 1.

We should probably grab some `semaphore` before checking term and index.

gleb-cloudius replied:
The spec says:

```
def fence(v):
   set the minimal accepted token metadata version of the fence to v, \
        unless the fence already rejects v
```

So shouldn't we just ignore the update that tries to move the fence back?

gusev-p replied:
> So shouldn't we just ignore the update that tries to move the fence back?

How to guarantee atomicity between these two lines:
```
co_await update_fence_version(version);
co_await _sys_ks.local().update_topology_fence_version(version);
```

gleb-cloudius replied:
> > So shouldn't we just ignore the update that tries to move the fence back?
> 
> How to guarantee atomicity between these two lines:
> 
> ```
> co_await update_fence_version(version);
> co_await _sys_ks.local().update_topology_fence_version(version);
> ```

To not override newer version with the older one in the persistent storage we can use timestamps. The newer write will use larger timestamp. It would have been easy if we would use timed uuid as the version identifier, but I think we can have a way to generate monotonic timestamp without it as well.

gusev-p replied:
> To not override newer version with the older one in the persistent storage we can use timestamps.

I’m not sure I understand. Are you proposing some checks in CQL, like `update scylla_local set value = x where key = y and value value < x`? Now it’s `set_scylla_local_param_as`, we will need to refactor it. 

A lock looks simpler:
```
case raft_topology_cmd::command::fence: {
    // just a single line, update_fence_version throws if version < current_version
    auto operation_holder = co_await get_units(_raft_topology_cmd_handler_state._operation_mutex, 1);

    const auto version = _topology_state_machine._topology.version;
    co_await update_fence_version(version);
    co_await _sys_ks.local().update_topology_fence_version(version);
    result.status = raft_topology_cmd_result::command_status::success;
    break;
}
```

gleb-cloudius replied:
> > To not override newer version with the older one in the persistent storage we can use timestamps.
> 
> I’m not sure I understand. Are you proposing some checks in CQL, like `update scylla_local set value = x where key = y and value value < x`? Now it’s `set_scylla_local_param_as`, we will need to refactor it.
> 

No. I propose to do `UPDATE scylla_local ... USING TIMESTAMP X` where a write with smaller version will have smaller timestamp. The visible write is the one with the newest timestamp.
 
> A lock looks simpler:
> 
> ```
> case raft_topology_cmd::command::fence: {
>     // just a single line, update_fence_version throws if version < current_version
>     auto operation_holder = co_await get_units(_raft_topology_cmd_handler_state._operation_mutex, 1);
> 
>     const auto version = _topology_state_machine._topology.version;
>     co_await update_fence_version(version);
>     co_await _sys_ks.local().update_topology_fence_version(version);
>     result.status = raft_topology_cmd_result::command_status::success;
>     break;
> }
> ```

A lock may work too, but timestamp is the usual way it is done lockless in Cassandra/Scylla.

kbr-scylla commented on service/storage_proxy.cc:
```diff
@@ -683,23 +677,7 @@ class storage_proxy::remote {
-        if (p->features().typed_errors_in_read_rpc) {
-            replica::exception_variant ex = replica::try_encode_replica_exception(eptr);
-            if (ex) {
-                co_return final_tuple_type(std::tuple(query::result_digest(), api::missing_timestamp, cache_temperature::invalid(), replica::exception_variant(std::move(ex)), std::nullopt));
```
The new code will return `final_tuple_type{}`, I see that the `long` will be different (it won't be `api::missing_timestamp` but `0`) - won't that be a problem?

gusev-p replied:
> won't that be a problem?

no, these values will be just dropped on the call site in case of exception, e.g.:
```
auto&& [result, hit_rate, opt_exception] = co_await ser::storage_proxy_rpc_verbs::send_read_data(&_ms, addr, timeout, cmd, pr, digest_algo, rate_limit_info, fence);
if (opt_exception.has_value() && *opt_exception) {
    co_await coroutine::return_exception_ptr((*opt_exception).into_exception_ptr());
}
```

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -985,6 +985,20 @@ class topology_coordinator {
+            raft_topology_cmd { raft_topology_cmd::command::barrier_and_drain },
+            true);
+        if (!exec_command_res) {
+            co_return std::make_pair(std::move(node), exec_command_res);
```
nit: `std::make_pair` is not necessary in new C++ versions, it was a workaround for a language defect, today we can use the constructor directly (`std::pair{...}`).

gusev-p commented on locator/token_metadata.hh:
```diff
@@ -309,6 +315,8 @@ mutable_token_metadata_ptr make_token_metadata_ptr(Args... args) {
 class shared_token_metadata {
     mutable_token_metadata_ptr _shared;
     token_metadata_lock_func _lock_func;
+    utils::phased_barrier _versions_barrier;
```
Do you really see value in such comments? The facts you mentioned (there's at most one operation in each version at any moment; we start it immediately when we set a new token metadata version and it's destroyed when that token metadata instance is destroyed.) can be deduced in a few seconds by just grepping `_versions_barrier` and finding `advance_and_await` and `start` both in method `shared_token_metadata::set`. It's faster than reading a comment.

kostja replied:
A comment  is also a manifest of the intent. You can't grep for what was on your mind when you wrote this code unless you write it down as well.

kbr-scylla replied:
> can be deduced in a few seconds

Maybe you're much smarter than me, but it took me at least a few minutes to understand what's happening

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5064,6 +5069,18 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+        {
+            auto& state = _raft_topology_cmd_handler_state;
+            if (state.term != term) {
+                state.term = term;
```
We're not using `state.term` for anything?

gusev-p replied:
`last_index` works only in the context of some `term`, if the `term` changes (coordinator failover took place) we just update the `term` and `last_index` without checking the `last_index`.

kbr-scylla replied:
Right - we use it because of the `else if`.

kostja replied:
Actually we never send uncommitted commands. For committed commands the term never changes. So it should be sufficient to track cmd_index, this will guarantee we never apply the command twice or from an outdated coordinator. What concerns me is that we can miss gaps in cmd_index, e..g due to a software bug we will skip some commands on the follower without noticing. next/prev state id approach guarantees this doesn't happen. but this is outside the scope of this patch.

kbr-scylla commented on docs/dev/raft-topology.md:
```diff
@@ -0,0 +1,115 @@
+Topology change algorithm over raft with failure recovery
```
If you're writing python-like pseudocode, and using "```python ... ```" blocks for syntax coloring, then please use the right comment markers (I see sometimes there's #, sometime's there's //, let's use # consistently), and replace `version++` with `version += 1`

kbr-scylla commented on locator/token_metadata.cc:
```diff
@@ -73,6 +73,7 @@ class token_metadata_impl final {
     static thread_local long _static_ring_version;
 
     token_metadata::version_t _version = 0;
+    token_metadata::version_tracker_t _version_tracker;
```
It will be lost in the depths of the git log

gusev-p commented:
new version:
  * addressed all the comments, [the changes](https://github.com/scylladb/scylladb/compare/98cc500cbd0aadde2996c784e11e41abcf0b4c23..afa346c5d9479d8d4b91912d1686e5f9e53dd02f) 

@kbr-scylla

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1549/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1566/

kbr-scylla commented:
@gusev-p please rebase

gusev-p commented:
* rebased on master
* dropped `test_tablets.py: execute -> run_async` commit

kbr-scylla commented on locator/token_metadata.hh:
```diff
@@ -310,13 +316,32 @@ class shared_token_metadata {
+    //   * A new phase begins when a mutable_token_metadata_ptr passed to shared_token_metadata::set has
+    //   a higher version than the current one.
+    //   * Each shared_token_metadata::set call initiates an operation on the barrier. If multiple calls
+    //   have the same version, multiple operations may be initiated with the same phase.
```
In what situation will there be multiple calls to `shared_token_metadata::set` with the same version?

gusev-p replied:
There is no requirement that every change of `token_metadata` should increment the version or be followed by `barrier_and_wait`, for some stages of topology change sm this can be skipped for efficiency.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1628/

gusev-p commented on service/storage_service.cc:
```diff
@@ -4905,6 +4905,21 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+                    co_await container().invoke_on_all([version] (storage_service& ss) -> future<> {
+                        slogger.debug("Got raft_topology_cmd::barrier_and_drain, version {}", version);
+                        if (const auto v = ss._shared_token_metadata.get()->get_version(); v != version) {
+                            on_internal_error(slogger,
```
fixed

gusev-p commented on service/storage_service.cc:
```diff
@@ -5057,13 +5057,32 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+                    auto holder = co_await get_units(_raft_topology_cmd_handler_state._operation_mutex, 1);
+                    const auto version = _topology_state_machine._topology.version;
+                    co_await update_fence_version(version);
+                    co_await _sys_ks.local().update_topology_fence_version(version);
```
no, `storage_service::update_fence_version` updates only in-memory fence version, we don't need lock and updating on-disk version in `main.cc` where `storage_service::update_fence_version` is used second time.

gusev-p commented on service/storage_proxy.cc:
```diff
@@ -2811,6 +2811,34 @@ storage_proxy::mutate_hint(const schema_ptr& s, const frozen_mutation& m, tracin
+    if (!token || token.topology_version >= fence_version) {
+        return std::nullopt;
+    }
+    static logger::rate_limit rate_limit(std::chrono::seconds(1));
```
fixed

gusev-p commented on service/storage_proxy.cc:
```diff
@@ -498,22 +505,25 @@ class storage_proxy::remote {
-                        clock_type::time_point timeout) {
-                    return p->mutate_locally(std::move(s), m, std::move(tr_state), db::commitlog::force_sync::no, timeout, smp_grp, rate_limit_info);
+                        clock_type::time_point timeout, fencing_token fence) {
+                    return p->apply_fence(p->mutate_locally(std::move(s), m, std::move(tr_state), db::commitlog::force_sync::no, timeout, smp_grp, rate_limit_info), fence);
```
The first `apply_fence` is in `handle_write` to reuse the error handling logic, we need to make a proper call to `send_mutation_failed`. The second `apply_fence` is inside `apply_fn` for consistency with all the other local `storage_service` calls that need to be fenced (`apply_locally`, `query_mutations_locally`, `query_result_local`, `query_result_local_digest`), their futures are all just wrapped in `apply_fence` in a similar manner.

gusev-p commented:
new version:
  * fixed @kbr-scylla comments, changes https://github.com/scylladb/scylladb/compare/0dca2627b029bbd621042a8bfdba7f34d463ed27..0180a6e1947e3c99cb1a7254cf017417c31fd353

kbr-scylla commented:
I'll queue once CI passes

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1639/

kbr-scylla commented:
> I'll queue once CI passes

@kostja and @tgrabiec wanted to review first, let's wait.

kostja commented on service/storage_service.cc:
```diff
@@ -4921,6 +4926,18 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+            if (state.term != term) {
+                state.term = term;
+            } else if (cmd_index <= state.last_index) {
+                // Return an error since the command is outdated
```
Actually we never send uncommitted commands. For committed commands the term never changes. So it should be sufficient to track cmd_index, this will guarantee we never apply the command twice or from an outdated coordinator. What concerns me is that we can miss gaps in cmd_index, e..g due to a software bug we will skip some commands on the follower without noticing. next/prev state id approach guarantees this doesn't happen. but this is outside the scope of this patch.

gleb-cloudius replied:
This has nothing to do with raft commands. This is about RPCs from the coordinator.

kostja replied:
OK, I realized this is a volatile state. OK, so we have raft commit index, state id and now a volatile cmd index which happens to have the same name as raft command index but stands for something different. I hope you see my point: the coordinator state is uniquely identified by raft commit index. It could have been used instead.

gleb-cloudius replied:
It cannot. First of all I do not think we should expose it, but more important is that for each raft index we may have multiple rpcs, so it is not unique.

tgrabiec commented on service/storage_service.cc:
```diff
@@ -998,6 +998,20 @@ class topology_coordinator {
 
+    future<std::pair<node_to_work_on, bool>> global_token_metadata_barrier(node_to_work_on&& node) {
+        bool exec_command_res;
+        std::tie(node, exec_command_res) = co_await exec_global_command(std::move(node),
```
`exec_global_command()` only works on normal nodes, but we want the barrier to work with all members.

gleb-cloudius replied:
Why? None normal nodes do not suppose to participate in queries.

tgrabiec replied:
Bootstrapping node participates as a replica. So we want fence to take effect on them.

gleb-cloudius replied:
But not as a coordinator and we drain coordinator requests here, not replicas.

tgrabiec replied:
Ok, but what about the next invocation, which sends the fence command?

gleb-cloudius replied:
The fence is needed to fail commands that we did not managed to drain. But non of them were for the bootstrapping node. First barrier drains writes that goes to old nodes only, second drains reads that goes throw the old nodes.

tgrabiec replied:
Ok, it seems to work given current algorithm. It may break if we add other kinds of modes which rely on the promise (which we don't provide) that after the barrier only requests using latest topology version will be admitted to any node.

tgrabiec commented on service/storage_service.cc:
```diff
@@ -679,7 +679,8 @@ class topology_coordinator {
     raft::server& _raft;
-    raft::term_t _term;
+    const raft::term_t _term;
+    size_t _last_cmd_index = 0;
```
Isn't uint64_t more appropriate? The index is not tied to memory size.

tgrabiec commented on service/storage_service.cc:
```diff
@@ -840,19 +841,21 @@ class topology_coordinator {
         if (!ip) {
-            slogger.info("raft topology: cannot send command {} to {} because mapping to ip is not available",
-                         cmd.cmd, id);
+            slogger.info("raft topology: cannot send command {} with term {} and index {} "
```
Shouldn't that be at least "warn"?

gusev-p replied:
maybe, @gleb-cloudius ?

gleb-cloudius replied:
Yes, it should higher severity than info. More importantly the id->ip mapping should not be here at all. but that's another story.

gusev-p replied:
ok, will update it

tgrabiec commented on service/storage_service.cc:
```diff
@@ -4921,6 +4926,18 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+        {
+            auto& state = _raft_topology_cmd_handler_state;
+            if (state.term != term) {
+                state.term = term;
```
@gusev-p How do you know if this isn't the old term? In which case, reordering will not be detected.

gleb-cloudius replied:
Because earlier we have:

```
            // do barrier to make sure we always see the latest topology
            co_await raft_server.read_barrier(&_abort_source);
            if (raft_server.get_current_term() != term) {
               // Return an error since the command is from outdated leader
               co_return result;
            }
```

tgrabiec replied:
Right

tgrabiec commented on idl/storage_service.idl.hh:
```diff
@@ -36,6 +36,6 @@ struct raft_topology_snapshot {
 
 struct raft_topology_pull_params {};
 
-verb raft_topology_cmd (raft::term_t term, service::raft_topology_cmd) -> service::raft_topology_cmd_result;
```
The commit log is missing explanation why do we need to prevent RPC reordering.

gusev-p replied:
fixed

tgrabiec commented on docs/dev/raft-topology.md:
```diff
@@ -0,0 +1,115 @@
+        # No uncompleted unfenced writes to only old replica set are in flight
+        stream()
+    except:
+        # If we are hear we did not manage to stream yet and thus should go back to old topology
```
hear -> here

gusev-p replied:
done

tgrabiec commented on docs/dev/raft-topology.md:
```diff
@@ -0,0 +1,115 @@
+        # a dead node, but may be those nodes are accessible by other nodes in the cluster that may
+        # still read from them let's try to isolate them and continue
+        Try:
+            isolate(dead nodes from the set of nodes that are losing range) # what to do if this also fails?
```
We cannot decide to isolate nodes because we don't know if they're permanently dead. Only admin can do this.

gleb-cloudius replied:
'Isolation' is not permanent. The node can rejoin the cluster after it syncs the topology. The doc omits how exactly isolation should be implemented. It cannot be done for all verbs on messaging level since we want raft to work to be able to catch up with the topology.

gleb-cloudius replied:
> Doesn't it mean that a temporary failure to communicate with them will turn it into a permanent failure which requires admin intervention to fix? This can impact cluster availability.

Not necessary. Are node may see that it was isolated (this will be a flag in raft) and do recovery automatically. But this series has nothing to do with it. This should be discussed on the spec the test from copied from.

tgrabiec replied:
Maybe the PR shouldn't document algorithm which isn't implemented yet.

gusev-p commented on service/storage_service.cc:
```diff
@@ -4905,6 +4905,20 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+                        slogger.debug("Got raft_topology_cmd::barrier_and_drain, version {}, current version {}",
+                            version, current_version);
+                        if (current_version == version) {
+                            co_await ss._shared_token_metadata.stale_versions_in_use();
```
> Why aren't we waiting for stale versions if current_version == version? Couldn't this violate the post-condition of barrier_and_drain which says that there are no stale requests?

The condition `current_version == version` can be violated only if the topology change coordinator changed (moved to another node) during `barrier_and_drain` and managed to increment version concurrently. In this case the current `barrier_and_drain` is moot and we can skip all the waiting. `stale_versions_in_use()` waits for all the the previous versions, this will be done in the `barrier_and_drain` RPC from the new coordinator.

tgrabiec replied:
So we can skip the waiting, because there shouldn't be anyone waiting for the result, if the system works as intended. But isn't it better to fail in this case? This is also correct, and more in line with the intent of the barrier. If our assumption don't hold for some reason, we will learn about it quickly rather than masking the problem with wrong behavior.

gusev-p replied:
> So we can skip the waiting, because there shouldn't be anyone waiting for the result, if the system works as intended. 

Yes.

> If our assumption don't hold for some reason, we will learn about it quickly rather than masking the problem with wrong behavior.

Makes sense, ok, will update.

gusev-p commented on service/storage_service.cc:
```diff
@@ -998,6 +998,20 @@ class topology_coordinator {
+            raft_topology_cmd { raft_topology_cmd::command::barrier_and_drain },
+            true);
+        if (!exec_command_res) {
+            co_return std::pair{std::move(node), exec_command_res};
```
I don't like exceptions in general since they complicate control flow. The additional channel for returning results from functions complicate matters, it's difficult to reason about it, whether we covered all the potential exception types in the corresponding catch on not.

In this example, the only difference will be `if` replaced with `try/catch`. I would like to hear from @gleb-cloudius, since he is the author of this code.

gleb-cloudius replied:
> Not introduced by this patch, but `exec_global_command_helper()` should not return a bool, but an exception. Returning bool will make debugging issues much harder, since you have to dig the error from the logs on a different node, and you won't know which error belongs to which failure.

I do not understand how throwing an exception will make any difference. And as we learned exceptions should not be used for flow control.

tgrabiec replied:
> In this example, the only difference will be if replaced with try/catch. I would like to hear from @gleb-cloudius, since he is the author of this code.

You can keep the `if`, just check `f.failed()` and throw or log `f.get_exception()`.

> I do not understand how throwing an exception will make any difference. 

The caller of global_token_metadata_barrier() will get the cause of exec_global_command() failure, so instead of logging "global barrier failed" it will log "global barrier failed: node 127.0.0.2: std::bad_alloc()"

> And as we learned exceptions should not be used for flow control.

Because of performance?

gleb-cloudius replied:
> > In this example, the only difference will be if replaced with try/catch. I would like to hear from @gleb-cloudius, since he is the author of this code.
> 
> You can keep the `if`, just check `f.failed()` and throw or log `f.get_exception()`.

That what `exec_global_command_helper` does exactly, no?


> 
> > I do not understand how throwing an exception will make any difference.
> 
> The caller of global_token_metadata_barrier() will get the cause of exec_global_command() failure, so instead of logging "global barrier failed" it will log "global barrier failed: node 127.0.0.2: std::bad_alloc()"

OK, but the error is already logged by whoever returned the failure.

> 
> > And as we learned exceptions should not be used for flow control.
> 
> Because of performance?

Yes.

tgrabiec replied:
> > > In this example, the only difference will be if replaced with try/catch. I would like to hear from @gleb-cloudius, since he is the author of this code.
> > 
> > 
> > You can keep the `if`, just check `f.failed()` and throw or log `f.get_exception()`.
> 
> That what `exec_global_command_helper` does exactly, no?

Yes. The point was that we don't need to add try/catch when we work with exceptions, which was brought up by @gusev-p.

> 
> > > I do not understand how throwing an exception will make any difference.
> > 
> > 
> > The caller of global_token_metadata_barrier() will get the cause of exec_global_command() failure, so instead of logging "global barrier failed" it will log "global barrier failed: node 127.0.0.2: std::bad_alloc()"
> 
> OK, but the error is already logged by whoever returned the failure.

So the information is probably there. But if you were to compose the failure report in a hierarchy, and log that a higher operation failed, it makes it hard to analyze the failure if the cause is in a different line.  Also, information about causality may be lost this way. For example, are you sure that the failure you see on the cmd level belongs is the failure on a higher level, or maybe it belongs to an earlier retry which eventually succeed. I can see there is no such logging now, but we probably will want it at some point.

> 
> > > And as we learned exceptions should not be used for flow control.
> > 
> > 
> > Because of performance?
> 
> Yes.

In this case, it should be rare. We should add sleep between failure and retry so that we don't retry+fail in a loop which is too tight.

gusev-p replied:
> So the information is probably there. But if you were to compose the failure report in a hierarchy, and log that a higher operation failed, it makes it hard to analyze the failure if the cause is in a different line. Also, information about causality may be lost this way. For example, are you sure that the failure you see on the cmd helper level is the cause of a failure on a higher level, or maybe it belongs to an earlier retry which eventually succeed. I can see there is no such logging now, but we probably will want it at some point.

I added [a commit](https://github.com/scylladb/scylladb/pull/13884/commits/b4d0025e10c1dd1df22d83a3a45202eef85b571a) with the refactoring.

> In this case, it should be rare. We should add sleep between failure and retry so that we don't retry+fail in a loop which is too tight.

Let's postpone this until the proper error handling implementation.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -5056,13 +5056,32 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+                    // checks that the version doesn't decrease, we do the check and persist
+                    // the new version under the same lock to avoid raises.
+                    auto holder = co_await get_units(_raft_topology_cmd_handler_state._operation_mutex, 1);
+                    const auto version = _topology_state_machine._topology.version;
```
We make sure that we do not execute commands out of order. It is better to do it once then verify each command separately.

tgrabiec replied:
I see. read_barrier() guarantees that we see state which is at least as recent than on the caller of the barrier. What if there was a failover after we checked the term and index but before we acquired the _operation_mutex? We may bring the fence on a different (later) version than intended.

gusev-p replied:
> What if there was a failover after we checked the term and index but before we acquired the _operation_mutex? We may bring the fence on a different (later) version than intended.

Yes, it's a bug, we should capture the version immediately after the checks, before any yields, will update it.

gusev-p commented:
next version:
  * `cmd_index: size_t -> uint64_t`;
  * log caller ip if `apple_fence` failed;
  * [storage_service: warn if can't find ip for server](https://github.com/scylladb/scylladb/pull/13884/commits/0cde8b265754e7294d76e9ec82d119c7fff1f31d)
  * comment `fencing_token`;
  * `cmd_index` and raft command reordering, elaborated on motivation in the commit message. 

changes: https://github.com/scylladb/scylladb/compare/0180a6e1947e3c99cb1a7254cf017417c31fd353..0cde8b265754e7294d76e9ec82d119c7fff1f31d

avikivity commented:
> This is the initial implementation of [this spec](https://docs.google.com/document/d/1X6pARlxOy6KRQ32JN8yiGsnWA9Dwqnhtk7kMDo8m9pI/edit).

> * when the version is incremented, a `barrier_and_drain` command is sent to all the nodes in the cluster, if some node is unavailable we fail and retry indefinitely;

What if the node is unavailable indefinitely?

gleb-cloudius commented:
> > This is the initial implementation of [this spec](https://docs.google.com/document/d/1X6pARlxOy6KRQ32JN8yiGsnWA9Dwqnhtk7kMDo8m9pI/edit).
> 
> > * when the version is incremented, a `barrier_and_drain` command is sent to all the nodes in the cluster, if some node is unavailable we fail and retry indefinitely;
> 
> What if the node is unavailable indefinitely?

There is no error handling in the coordinator now and the patch adds fencing, not error handling. We will handle error on top of this series.

avikivity commented:
> > > This is the initial implementation of [this spec](https://docs.google.com/document/d/1X6pARlxOy6KRQ32JN8yiGsnWA9Dwqnhtk7kMDo8m9pI/edit).
> > 
> > 
> > > * when the version is incremented, a `barrier_and_drain` command is sent to all the nodes in the cluster, if some node is unavailable we fail and retry indefinitely;
> > 
> > 
> > What if the node is unavailable indefinitely?
> 
> There is no error handling in the coordinator now and the patch adds fencing, not error handling. We will handle error on top of this series.

Ok. @gusev-p would be nice to note in the cover letter.

gusev-p commented:
> Ok. @gusev-p would be nice to note in the cover letter.

It's already noted:

> along with fencing, the spec above also describes error handling, isolation and --ignore_dead_nodes parameter handling, these will be also added later;

I created the [ticket](https://github.com/scylladb/scylladb/issues/14070) with todo items.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1828/

gusev-p commented:
new version:
  * rebased on master
  * dropped `raft-topology.md` (this document is not ready, it should be rewritten in more detail when all the fencing parts are discussed, stabilised and implemented, I added an item to the https://github.com/scylladb/scylladb/issues/14070);
  * [storage_service: exec_global_command: bool result -> exceptions](https://github.com/scylladb/scylladb/pull/13884/commits/d9d29ec293f039a0ecffdda4c94150323e71c753) refactoring;
  * fix the bug with the fencing command handler, the current version may change, now it's captured right after the checks, before any yields;
  * `barrier_and_drain` command handler fix, throw exception if the current version on a shard doesn't match the initial version.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1849/

gusev-p commented:
@tgrabiec the tests passed, please merge

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -1121,20 +1125,27 @@ class topology_coordinator {
+                                      "raft_topology_cmd::command::barrier failed, error {}",
+                                      f.get_exception());
+                        break;
+                    }
```
Missing `node = std::move(f).get()`?
But I wonder how is it possible that the tests passed with this missing...

gusev-p replied:
> But I wonder how is it possible that the tests passed with this missing...

`node.rs->state == node_state::removing` part of the `if` which is next to this code just executed `raft_topology_cmd::command::stream_ranges` without guard, the `else` part does `release_node` at the beginning of `exec_direct_command`.

I'll fix this in the follow-up.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5078,13 +5084,31 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+                    // checks that the version doesn't decrease, we do the check and persist
+                    // the new version under the same lock to avoid raises.
+                    auto holder = co_await get_units(_raft_topology_cmd_handler_state._operation_mutex, 1);
+                    co_await update_fence_version(version);
```
Since `version` is now defined above, before `get_units` which may yield, and the topology version may change during the yield`, `update_fence_version` may actually detect a mismatch, IIUC.

That's good, but it's no good that we use `on_internal_error` for a situation which is not a bug.
We should either update `update_fence_version` to not use `on_internal_error` or do another check here. If the check fails we should use a regular throw since it's a situation that may arise, not a bug.

gusev-p replied:
Fixed [here](https://github.com/gusev-p/scylla/tree/raft_fencing3_fix), will turn it to PR when this is merged.
